### ISSUE TITLE
Notes #1

### ISSUE THREAD

**Issue by PhantomWatson** [BallStateCBER/whyarewehere#8]
Users
------
- [x] Since administrators create user accounts, there doesn't need to be a page where the general public can create their own accounts
- [x] `/src/Template/Users/account.ctp` lets anyone change their role to an administrator
- [x] I don't see any means for administrators to add or edit user accounts
- [x] Speaking of which, `/src/Template/Users/index.ctp` has a link to the page `/users/add`, which hasn't been created
- [x] I see the field `Users.end_date` in the database, but no means of setting it on the website
- [x] Clicking on "Hi!" to edit your account information isn't intuitive
- [x] I'm not sure what the function of the new "employee" role is, as distinct from administrators. The requirements only call for student and admin roles, but if you're shooting for being able to distinguish administrators who supervise students from those who don't (e.g. to narrow down a select field of possible supervisors), a `Users.is_supervisor` boolean flag would be way simpler and wouldn't complicate any of the role/authorization code.
- [x] `/register` is accessible to the public and allows people to create admin accounts for themselves
- [x] Home page still refers to creating your own account
- [x] On `/employees`, when no position is specified, the title line reads "Current ." or "Former ."

Validation / Authorization
----------------------------
- [x] I'm not sure that it's okay for us to release any of this collected information to the public (and some stuff like phone numbers and emergency contacts should definitely be hidden), so for now the only public-facing pages should be a basic home page and a login page.
- [x] I don't see a validation rule for "Students should not be able to create a new work report if another one exists for the same student and same project without an end date"
- [x] I don't see a validation rule for "Students should not be able to create a new work report with a start date that’s between the start and end date of an existing work report matching the same student and project"
- [x] For safety's sake, it would be good to prevent non-admins from being able to edit or delete records that they're not associated with. I suggest authorization work like this:
   - admins -> everything
   - students -> only their own work reports and account info
- [x] Links to inaccessible sites still in header when not logged in

Work reports
---------------
- [x] The field in `/reports/add` for specifying the student associated with a work report is disabled
- [x] The option to not specify an end date is not provided, and the validation rules in `ReportsTable` make it required
- [x] I don't see where these requirements are met:
   - View student work
      - Current (work by current students with end date > today or null)
      - Past work
      - Associated with a specific student
      - Associated with a specific project
 - [x] When I click on "Reports by project name", I see "Notice (8): Trying to get property of non-object [APP/Template\Element\reports\index.ctp, line 85]"

Projects
---------
- [x] Not all projects assigned to students can be categorized as either community projects, publications, or websites. Events and unpublished research, for instance. Categorizing projects isn't necessary for our requirements, so the problem could be fixed by removing the concept of categories altogether, adding a catch-all "other projects" category, and/or talking to the student supervisors and collecting all of the categories that they expect projects to fall under.
- [x] Projects are missing an optional "fund number" field
- [x] It appears that there's no way to add a project and _not_ specify a grant
- [x] I don't see a field for projects that fulfills this requirement: "Multi-line text field for funding details (e.g. $5k for undergraduate student work, $10k for graduate student work, etc.)"
- [x] The flow for administrators to log in and add or edit projects isn't clear to me. The index pages for the different project types only display "add" links if no matching projects exist. Ideally, there's a `/projects` index page with add, edit, delete, view links.
- [x] The field Projects.grant_name is missing

Miscellaneous
----------------
- [x] "Go ahead and _____ so this lame placeholder message goes away." is **very ableist language**  😱
- [x] I recommend adding an "administrator's email address" value to `Configure` so it's defined in one place instead of hard-coding your email address in multiple places.

**Comment by ericadestoya** [comment]
Erg, "lame" is my hardest one to break as far as the ableist language goes! Call-out received, this list will be done hopefully by today. :)

**Comment by ericadestoya** [comment]
Note: I had changed user registration to be admin-only as per this list, but in retrospect I think students should be able to add themselves without needing an admin to do that for them. That way they can get started right away.

**Comment by ericadestoya** [comment]
So, now all that's left to do are the reports indexing specs (view by students, sups, projects, etc) and the auth stuff. Of course I will go through auth stuff with a finetoothed comb. However, I was wondering: is the goal of "View student work" to _order_ student work by these different things, or is it to _only show_ student work with these properties? As of now, our paginator can _order_ student work by their projects, people involved, dates, etc. but cannot yet _only show_ certain projects. If the objective is to _only show_, let me know because I'd been wanting to build a search feature anyway.

**Comment by PhantomWatson** [comment]
Filtering, rather than ordering. I suggest these pages:

`/reports`: View all work reports
`/reports/current`: View only current (uncompleted) work reports

    $query->where([
        'OR' => [
            function ($exp, $q) {
                return $exp->isNull('end_date');
            },
            function ($exp, $q) {
                return $exp->gte('end_date', date('Y-m-d H:i:s'));
            },
    ])

`/reports/past`: View only past (completed) work reports

    $query->where([
        function ($exp, $q) {
            return $exp->lt('end_date', date('Y-m-d H:i:s'));
        }
    ])

`/reports/student/:student-id`: Associated with a specific student
`/reports/supervisor/:supervisor-id`: Associated with a specific student
`/reports/project/:project-id`: Associated with a specific project

**Comment by ericadestoya** [comment]
Cool, can do. Filtering it is!

**Comment by PhantomWatson** [comment]
I've unchecked the following:
 - "The field in /reports/add for specifying the student associated with a work report is disabled"...
   _(It's still disabled)_
 - The flow for administrators to log in and add or edit projects isn't clear to me...
   _(In /projects, I don't see any add/edit/delete links)_
 - I don't see a field for projects that fulfills this requirement: "Multi-line text field for funding details..."
   _(I see `funds.funding_details` was added, but Srikant suggests that we'll need a field in `projects` to record information about how money is divided up in specific projects)_

I've added the following:
 - `/register` is accessible to the public...
 - Home page still refers to...
 - Links to inaccessible sites...
 - On `/employees`, when no position is specified...
 - When I click on "Reports by project name"...
 - The field Projects.grant_name is missing

**Comment by ericadestoya** [comment]
`/register` is accessible to the public in order for students to be able to register themselves without needing an admin to do it for them. I could add some other validation measures, like sending a confirmation email to their @bsu.edu email. Also, only admins can add other admins. I could make `/register` an admin-only page altogether, but I feel like that's gonna eat into our time.

The student field in `/reports/add` is now disabled for everyone _except admins_. This is so that students can only report on their own work. Admins can do whatever they want in `/reports/add`. Is this a good solution?

Position issue in `/employees` resolved.

`Reports by name` issue resolved.

**Comment by ericadestoya** [comment]
`/projects` now shows edit links for admins. Add links are available at the header always. I'm just trying to make the edit links for projects, users, staff etc. visible only to admins.

**Comment by PhantomWatson** [comment]
> I could add some other validation measures, like sending a confirmation email

Confirmation emails ain't a bad idea, but probably aren't necessary.

> Also, only admins can add other admins.

`/register` shows a role selector with "Site Admin" and "Student" as the options, though for reasons I don't understand, neither option actually gets recorded in the database. `role` just gets saved as `null`, but that doesn't seem to screw anything up, as I can't find anything that explicitly checks if the role === 'Student'.

**Comment by ericadestoya** [comment]
Whoops. Welp, I will be overhauling `/register` tomorrow then.

**Comment by PhantomWatson** [comment]
> I could make /register an admin-only page

Yep, that's the idea. At least until the project requirements change.

> The student field in /reports/add is now disabled for everyone except admins

Heads-up: [This line](https://github.com/BallStateCBER/whyarewehere/blob/master/src/Controller/ReportsController.php#L277) has the effect of ignoring what was entered and always making student_id the ID of the user.

> Is this a good solution?

Yep! I can't think of a use case where a non-admin would need to report work for someone else.

**Sidenote: I just noticed a new issue.**
 - Because `id` is a field in the registration form, it effectively makes it an _edit_ form if you enter someone else's ID. When I tried to make a new student account with the same fake BSU ID as an existing admin account, it actually overwrote the admin account.
 - If the site's being used properly, I wouldn't expect this to be an issue, but a mistyped student ID might cause chaos.
 - If we'd like to prevent this, the simplest solution would be making `Users.id` a simple auto-incremented key and adding `Users.bsu_id` to keep track of this data.

**Comment by ericadestoya** [comment]
>  - Because `id` is a field in the registration form, it effectively makes it an _edit_ form if you enter someone else's ID. When I tried to make a new student account with the same fake BSU ID as an existing admin account, it actually overwrote the admin account.

My solution:

```
 $oldUser = $this->Users->find()
    ->where(['id' => $user->id])
    ->first();
if (isset($oldUser->id)) {
    return $this->Flash->error('This user ID is already in use?');
}
```

**Comment by PhantomWatson** [comment]
Nice! But if you want to be extra efficient, you could just...

```
if (!$user->isNew()) {
    return $this->Flash->error('This user ID is already in use?');
}
```

**Comment by ericadestoya** [comment]
oml I can do that? Really? Doing that now. Then the whole `users` section should be done.

My main concern with making `/register` admin-only was that it would suck up admins' time (read: my time, your time, Srikant's time, etc) adding users. I don't mind doing the bulk of this, though.

**Comment by ericadestoya** [comment]
>  - I don't see a field for projects that fulfills this requirement: "Multi-line text field for funding details..."
   _(I see `funds.funding_details` was added, but Srikant suggests that we'll need a field in `projects` to record information about how money is divided up in specific projects)_

I'm apprehensive about putting any funding details in project pages. Admins can reference the fund number, and see details in the page for the fund, but if we've got funding details in the project section then it's no longer admin-only.

**Comment by PhantomWatson** [comment]
> I'm apprehensive about putting any funding details in project pages. Admins can reference the fund number, and see details in the page for the fund

Do you mean having `/projects/view/:id` exclude `Projects.fund_details` and instead having that field visible on `/funds/view/:id`, like in a list of details for all projects associated with a given fund?

> if we've got funding details in the project section then it's no longer admin-only.

Why won't project adding/editing be admin-only after we add the `fund_details` field?

**Comment by ericadestoya** [comment]
>> I'm apprehensive about putting any funding details in project pages. Admins can reference the fund number, and see details in the page for the fund

> Do you mean having `/projects/view/:id` exclude `Projects.fund_details` and instead having that field visible on `/funds/view/:id`, like in a list of details for all projects associated with a given fund?

Yep, exactly!

>> if we've got funding details in the project section then it's no longer admin-only.

> Why won't project adding/editing be admin-only after we add the `fund_details` field?

Students can add & edit (their own) projects. But I have it set up now so that everything to do with funding is admin-only. Students can only see fund numbers; however, I'd like to set it up so that admins can click on the fund numbers in the project pages and see the funding details. Should students have access to this information?

**Comment by ericadestoya** [comment]
Students now cannot create reports if a report exists for the same student & project and `end_date` is null or today or in the future. Basically they can't create duplicate work reports.

**Comment by PhantomWatson** [comment]
> Yep, exactly!

I'd do both, just to save people from having to click over to a second page to get the rest of the info for a project. Have all associated data from other models on `/funds/view/:id`, but _also_ have all fields from a project record visible on `/projects/view/:id`.

> Students can add & edit (their own) projects.

Whuh oh. Our requirements call for project adding/editing to be admin-only.

> Should students have access to this information?

Our requirements say that students should only be able to add and edit their own work reports and user info. The only information from other models that students need to interact with are `<select>` fields populated with existing projects/supervisors in work report forms.

**Comment by ericadestoya** [comment]
> I'd do both, just to save people from having to click over to a second page to get the rest of the info for a project. Have all associated data from other models on `/funds/view/:id`, but _also_ have all fields from a project record visible on `/projects/view/:id`.

Hmm... instead of having a whole new field in the database, I could just run a query that finds the `funding_details` for that particular fund and displays the details in a popover?

> Whuh oh. Our requirements call for project adding/editing to be admin-only.

Whoops. I'll fix that. Question: should students be able to _view_ projects?

**Comment by PhantomWatson** [comment]
> Hmm... instead of having a whole new field in the database, I could just run a query that finds the funding_details for that particular fund and displays the details in a popover?

Nope, the database structure that I sent you was set up that way _tooootally_ on purpose. Multiple projects that share the same `Projects.fund_number` can have different `Project.fund_details`. Like this:

```
Project 1 (The Culling)
Fund: 8675309 (General budget)
Details: $2,000 for rare African parrots, $30 for one hammer

Project 2 (The Cleanup)
Fund: 8675309 (General budget)
Details: $20 for a mop, $5 for bleach, $80 for empty pillows to put the feathers in
```

> should students be able to view projects?

There's no particular reason for students to view anything other than project names, so I wouldn't bother setting up student access to project funding info.

**Comment by ericadestoya** [comment]
Cool, I added `Projects.funding_details` and restricted students to just viewing & indexing Projects. Including removing links to reference adding & editing Projects, if one is a student.

**Comment by ericadestoya** [comment]
`Projects.grant_name` has been added in sql, field created in form, and testing expanded to include `grant_name`. For now I have closed the issue but I'll reopen if if I go through the list again and see that any of it is not implemented or sub-par.
